src/IssueViz.Security.UnitTests/ListenerFactoryTests.cs

```diff
@@ -0,0 +1,146 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Linq;
+using System.Net;
+using FluentAssertions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SonarLint.VisualStudio.Integration.UnitTests;
+using SonarLint.VisualStudio.IssueVisualization.Security.OpenInIDE;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.UnitTests
+{
+    [TestClass]
+    public class ListenerFactoryTests
+    {
+        [TestMethod]
+        [DataRow(10000, 10003, 10000)]
+        [DataRow(10000, 10003, 10001)]
+        [DataRow(10000, 10003, 10002)]
+        [DataRow(20000, 20003, 20003)]
+        [DataRow(20000, 20005, 20006)] // no free ports in the range
+        public void Create_FirstAvailablePortIsUsed(int startPort, int endPort, int firstFreePort)
+        {
+            IListenerFactory testSubject = new ListenerFactory(new TestLogger(logToConsole: true));
+
+            bool isFreePortInRange = firstFreePort >= startPort && firstFreePort <= endPort;
+
+            var allPorts = CreateRange(startPort, endPort);
+
+            // If any of the ports required by the test are not available then bail out.
+            if(!AreAllPortsAvailable(allPorts))
+            {
+                Assert.Inconclusive("Test setup error: some test ports are in use.");
+            }
+
+            // The unavailable ports are all at the start of the range.
+            var unavailablePorts = CreateRange(startPort, firstFreePort - 1);
+            var remainingFreePorts = CreateRange(firstFreePort + 1, endPort);
+
+            using(new PortGrabber(unavailablePorts))
+            using (var actual = testSubject.Create(startPort, endPort))
+            {
+                if (isFreePortInRange)
+                {
+                    var expectedPort = GetPrefix(firstFreePort);
+                    actual.Should().NotBeNull();
+                    actual.IsListening.Should().BeTrue();
+                    actual.Prefixes.Count.Should().Be(1);
+                    actual.Prefixes.First().Should().Be(expectedPort);
+                    IsPortAvailable(firstFreePort).Should().BeFalse();
+                }
+                else
+                {
+                    actual.Should().BeNull();
+                }
+
+                AreAllPortsAvailable(remainingFreePorts).Should().BeTrue();
+            }
+
+            // The port used by the test subject should have been released on disposal
+            IsPortAvailable(firstFreePort).Should().BeTrue();
+        }
+
+        private static int[] CreateRange (int start, int end)
+        {
+            if (start > end)
+            {
+                return Array.Empty<int>();
+            }
+            return Enumerable.Range(start, end - start + 1).ToArray();
+        }
+
+        private static bool AreAllPortsAvailable(params int[] ports) =>
```

Comment:
Cosmetics, maybe `VerifyTestPortsAreAvailable`? I can't put my finger on why but `AreAllPortsAvailable` is somehow harder to understand for me, not sure why 🤔

Reply:
The method is being called twice, first to check that all of the ports being used in the test are available, and then to check that the ports above the free port are still available.
I originally had a method with a name like `VerifyTestPortsAreAvailable`; it worked ok for the first case but didn't fit for the second.
Also, I deliberately moved the `Verify`/`Check` part out of the method as I want to take different actions. If all of the required ports are not free at the start then the test result should be Inconclusive rather than a failure. If the remaining ports aren't free after the listener is created then the test should fail.